libraries/standard/mqtt/src/iot_mqtt_subscription.c

```diff
@@ -88,108 +142,123 @@ static bool _packetMatch( const IotLink_t * pSubscriptionLink,
+    bool status = false, terminalChar = false;
 
-    /* Because this function is called from a container function, the given link
-     * must never be NULL. */
-    IotMqtt_Assert( pSubscriptionLink != NULL );
+    /* Determine if the terminal character is reached for both topic name and topic
+     * filter for the '#' wildcard. */
+    terminalChar = ( nameIndex == topicNameLength - 1 ) && ( filterIndex == topicFilterLength - 3 );
 
-    _mqttSubscription_t * pSubscription = IotLink_Container( _mqttSubscription_t,
-                                                             pSubscriptionLink,
-                                                             link );
-    _topicMatchParams_t * pParam = ( _topicMatchParams_t * ) pMatch;
+    if( terminalChar == true )
+    {
+        /* Determine if the topic filter ends with the '#' wildcard. */
+        status = ( pTopicFilter[ filterIndex + 1 ] == '/' ) && ( pTopicFilter[ filterIndex + 2 ] == '#' );
 
-    /* Extract the relevant strings and lengths from parameters. */
-    const char * pTopicName = pParam->pTopicName;
-    const char * pTopicFilter = pSubscription->pTopicFilter;
-    const uint16_t topicNameLength = pParam->topicNameLength;
-    const uint16_t topicFilterLength = pSubscription->topicFilterLength;
+        if( status == true )
+        {
+            goto cleanup;
+        }
+    }
 
-    /* Check for an exact match. */
-    if( topicNameLength == topicFilterLength )
-    {
-        status = ( strncmp( pTopicName, pTopicFilter, topicNameLength ) == 0 );
+    /* Determine if the terminal character is reached for both topic name and topic
+     * filter for the '+' wildcard. */
+    terminalChar = ( nameIndex == topicNameLength - 1 ) && ( filterIndex == topicFilterLength - 2 );
 
-        goto cleanup;
+    if( terminalChar == true )
+    {
+        /* Filter "sport/+" also matches the "sport/" but not "sport". */
+        status = ( pTopicFilter[ filterIndex + 1 ] == '+' );
     }
 
-    /* If the topic lengths are different but an exact match is required, return
-     * false. */
-    if( pParam->exactMatchOnly == true )
+cleanup:
+
+    return status;
+}
+
+/*-----------------------------------------------------------*/
+
+static bool _matchWildcards( const char * pTopicFilter,
+                             const char * pTopicName,
+                             uint16_t filterIndex,
+                             uint16_t topicNameLength,
+                             uint16_t * pNameIndex,
+                             bool * pMatch )
+{
+    bool status = false;
+
+    /* Check for wildcards. */
+    if( pTopicFilter[ filterIndex ] == '+' )
     {
-        status = false;
-        goto cleanup;
+        /* Move topic name index to the end of the current level.
+         * This is identified by '/'. */
+        while( ( *pNameIndex < topicNameLength ) && ( pTopicName[ *pNameIndex ] != '/' ) )
+        {
+            ( *pNameIndex )++;
+        }
+
+        ( *pNameIndex )--;
```

Comment:
Is there any chance that pNameIndex was zero?

Reply:
Yes, if comparing against the topic name '/'.  That would cause this line to decrement nameIndex to 65535.  But line 267 accounts for this and would bring it back to 0.  The end behavior is still correct: the topic match comparison would fail.